File: Sources/Ably/ARTLocalDeviceStorage.m
//
//  ARTLocalDeviceStorage.m
//  Ably
//
//  Created by Ricardo Pereira on 18/04/2017.
//  Copyright © 2017 Ably. All rights reserved.
//

#import "ARTLocalDeviceStorage.h"
#import "ARTLog.h"
#import "ARTLocalDevice+Private.h"

@implementation ARTLocalDeviceStorage {
    ARTLog *_logger;
}

- (instancetype)initWithLogger:(ARTLog *)logger {
    if (self = [super init]) {
        _logger = logger;
    }
    return self;
}

+ (instancetype)newWithLogger:(ARTLog *)logger {
    return [[self alloc] initWithLogger:logger];
}

- (nullable id)objectForKey:(NSString *)key {
    return [NSUserDefaults.standardUserDefaults objectForKey:key];
}

- (void)setObject:(nullable id)value forKey:(NSString *)key {
    [NSUserDefaults.standardUserDefaults setObject:value forKey:key];
}

- (NSString *)secretForDevice:(ARTDeviceId *)deviceId {
    NSError *error = nil;
    NSString *value = [self keychainGetPasswordForService:ARTDeviceSecretKey account:(NSString *)deviceId error:&error];

    if ([error code] == errSecItemNotFound) {
        [_logger debug:__FILE__ line:__LINE__ message:@"Device Secret not found"];
    }
    else if (error) {
        [_logger error:@"Device Secret couldn't be read (%@)", [error localizedDescription]];
    }

    return value;
}

- (void)setSecret:(NSString *)value forDevice:(ARTDeviceId *)deviceId {
    NSError *error = nil;
    if (value == nil) {
        [self keychainDeletePasswordForService:ARTDeviceSecretKey account:(NSString *)deviceId error:&error];

        if ([error code] == errSecItemNotFound) {
            [_logger warn:@"Device Secret can't be deleted because it doesn't exist"];
        }
        else if (error) {
            [_logger error:@"Device Secret couldn't be updated (%@)", [error localizedDescription]];
        }
    }
    else {
        [self keychainSetPassword:value forService:ARTDeviceSecretKey account:(NSString *)deviceId error:&error];

        if (error) {
            [_logger error:@"Device Secret couldn't be updated (%@)", [error localizedDescription]];
        }
    }
}

#pragma mark - Keychain

- (nonnull NSMutableDictionary *)newKeychainQueryForService:(nonnull NSString *)serviceName account:(nonnull NSString *)account {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionaryWithCapacity:3];
    [dictionary setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    [dictionary setObject:serviceName forKey:(__bridge id)kSecAttrService];
    [dictionary setObject:account forKey:(__bridge id)kSecAttrAccount];
    #if TARGET_OS_IPHONE
    [dictionary setObject:(__bridge id)kSecAttrAccessibleAlwaysThisDeviceOnly forKey:(__bridge id)kSecAttrAccessible];
    #endif
    return dictionary;
}

- (nonnull NSError *)keychainErrorWithCode:(OSStatus)status {
    NSString *message = nil;
    #if TARGET_OS_IPHONE
    switch (status) {
        case errSecUnimplemented: {
            message = @"errSecUnimplemented";
            break;
        }
        case errSecParam: {
            message = @"errSecParam";
            break;
        }
        case errSecAllocate: {
            message = @"errSecAllocate";
            break;
        }
        case errSecNotAvailable: {
            message = @"errSecNotAvailable";
            break;
        }
        case errSecDuplicateItem: {
            message = @"errSecDuplicateItem";
            break;
        }
        case errSecItemNotFound: {
            message = @"errSecItemNotFound";
            break;
        }
        case errSecInteractionNotAllowed: {
            message = @"errSecInteractionNotAllowed";
            break;
        }
        case errSecDecode: {
            message = @"errSecDecode";
            break;
        }
        case errSecAuthFailed: {
            message = @"errSecAuthFailed";
            break;
        }
        default: {
            message = @"errSecDefault";
        }
    }
    #else
    message = (__bridge_transfer NSString *)SecCopyErrorMessageString(status, NULL);
    #endif
    NSDictionary *userInfo = nil;
    if (message) {
        userInfo = @{ NSLocalizedDescriptionKey : message };
    }
    return [NSError errorWithDomain:[NSString stringWithFormat:@"%@.%@", ARTAblyErrorDomain, @"Keychain"] code:status userInfo:userInfo];
}

- (nullable NSString *)keychainGetPasswordForService:(nonnull NSString *)serviceName account:(nonnull NSString *)account error:(NSError *__autoreleasing *)error {
    NSMutableDictionary *query = [self newKeychainQueryForService:serviceName account:account];

    [query setObject:@YES forKey:(__bridge id)kSecReturnData];
    [query setObject:(__bridge id)kSecMatchLimitOne forKey:(__bridge id)kSecMatchLimit];
    CFTypeRef result = NULL;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &result);

    if (status != errSecSuccess) {
        if (error) {
            *error = [self keychainErrorWithCode:status];
        }
    }
    else {
        NSData *passwordData = (__bridge_transfer NSData *)result;
        if ([passwordData length]) {
            return [[NSString alloc] initWithData:passwordData encoding:NSUTF8StringEncoding];
        }
    }

    return nil;
}

- (BOOL)keychainDeletePasswordForService:(NSString *)serviceName account:(NSString *)account error:(NSError *__autoreleasing *)error {
    NSMutableDictionary *query = [self newKeychainQueryForService:serviceName account:account];
    OSStatus status;
    #if TARGET_OS_IPHONE
    status = SecItemDelete((__bridge CFDictionaryRef)query);
    #else
    CFTypeRef result = NULL;
    [query setObject:@YES forKey:(__bridge id)kSecReturnRef];
    status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &result);
    if (status == errSecSuccess) {
        status = SecKeychainItemDelete((SecKeychainItemRef)result);
        CFRelease(result);
    }
    #endif

    if (status != errSecSuccess && error != NULL) {
        *error = [self keychainErrorWithCode:status];
    }

    return (status == errSecSuccess);
}

- (BOOL)keychainSetPassword:(NSString *)password forService:(NSString *)serviceName account:(NSString *)account error:(NSError *__autoreleasing *)error {
    NSMutableDictionary *query = nil;
    NSMutableDictionary *searchQuery = [self newKeychainQueryForService:serviceName account:account];
    NSData *passwordData = [password dataUsingEncoding:NSUTF8StringEncoding];

    OSStatus status;
    status = SecItemCopyMatching((__bridge CFDictionaryRef)searchQuery, nil);
    if (status == errSecSuccess) { //item already exists, update it.
        query = [[NSMutableDictionary alloc] init];
        [query setObject:passwordData forKey:(__bridge id)kSecValueData];
        status = SecItemUpdate((__bridge CFDictionaryRef)(searchQuery), (__bridge CFDictionaryRef)(query));
    }
    else if (status == errSecItemNotFound) { //item not found, create it.
        query = [self newKeychainQueryForService:serviceName account:account];
        [query setObject:passwordData forKey:(__bridge id)kSecValueData];
        status = SecItemAdd((__bridge CFDictionaryRef)query, NULL);
    }

    if (status != errSecSuccess && error != NULL) {
        *error = [self keychainErrorWithCode:status];
    }

    return (status == errSecSuccess);
}

@end

File: Sources/AblySwift/ARTLocalDeviceStorage.swift
import Foundation
import Security

// swift-migration: original location ARTLocalDeviceStorage.h, line 9 and ARTLocalDeviceStorage.m, line 5
internal class ARTLocalDeviceStorage: NSObject, ARTDeviceStorage {
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 6
    private let logger: InternalLog
    
    // swift-migration: original location ARTLocalDeviceStorage.h, line 11 and ARTLocalDeviceStorage.m, line 9
    internal init(logger: InternalLog) {
        self.logger = logger
        super.init()
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.h, line 13 and ARTLocalDeviceStorage.m, line 16
    internal static func new(withLogger logger: InternalLog) -> ARTLocalDeviceStorage {
        return ARTLocalDeviceStorage(logger: logger)
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 20
    func objectForKey(_ key: String) -> Any? {
        return UserDefaults.standard.object(forKey: key)
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 24
    func setObject(_ value: Any?, forKey key: String) {
        UserDefaults.standard.set(value, forKey: key)
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 28
    func secretForDevice(_ deviceId: String) -> String? {
        do {
            let value = try keychainGetPassword(forService: ARTDeviceSecretKey, account: deviceId)
            return value
        } catch let error as NSError {
            if error.code == Int(errSecItemNotFound) {
                ARTLogDebug(logger, "Device Secret not found")
            } else {
                ARTLogError(logger, "Device Secret couldn't be read (\(error.localizedDescription))")
            }
            return nil
        }
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 42
    func setSecret(_ value: String?, forDevice deviceId: String) {
        do {
            if value == nil {
                try keychainDeletePassword(forService: ARTDeviceSecretKey, account: deviceId)
            } else {
                try keychainSetPassword(value!, forService: ARTDeviceSecretKey, account: deviceId)
            }
        } catch let error as NSError {
            if error.code == Int(errSecItemNotFound) {
                ARTLogWarn(logger, "Device Secret can't be deleted because it doesn't exist")
            } else {
                ARTLogError(logger, "Device Secret couldn't be updated (\(error.localizedDescription))")
            }
        }
    }
    
    // MARK: - Keychain
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 65
    private func newKeychainQuery(forService serviceName: String, account: String) -> [String: Any] {
        var dictionary: [String: Any] = [:]
        dictionary[kSecClass as String] = kSecClassGenericPassword
        dictionary[kSecAttrService as String] = serviceName
        dictionary[kSecAttrAccount as String] = account
        #if os(iOS) || os(watchOS) || os(tvOS)
        dictionary[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        #endif
        return dictionary
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 76
    private func keychainError(withCode status: OSStatus) -> NSError {
        var message: String?
        
        #if os(iOS) || os(watchOS) || os(tvOS)
        switch status {
        case errSecUnimplemented:
            message = "errSecUnimplemented"
        case errSecParam:
            message = "errSecParam"
        case errSecAllocate:
            message = "errSecAllocate"
        case errSecNotAvailable:
            message = "errSecNotAvailable"
        case errSecDuplicateItem:
            message = "errSecDuplicateItem"
        case errSecItemNotFound:
            message = "errSecItemNotFound"
        case errSecInteractionNotAllowed:
            message = "errSecInteractionNotAllowed"
        case errSecDecode:
            message = "errSecDecode"
        case errSecAuthFailed:
            message = "errSecAuthFailed"
        default:
            message = "errSecDefault"
        }
        #else
        message = SecCopyErrorMessageString(status, nil) as String?
        #endif
        
        var userInfo: [String: Any]? = nil
        if let message = message {
            userInfo = [NSLocalizedDescriptionKey: message]
        }
        
        return NSError(domain: "\(ARTAblyErrorDomain).Keychain", code: Int(status), userInfo: userInfo)
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 130
    private func keychainGetPassword(forService serviceName: String, account: String) throws -> String? {
        var query = newKeychainQuery(forService: serviceName, account: account)
        
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status != errSecSuccess {
            throw keychainError(withCode: status)
        } else {
            if let passwordData = result as? Data, !passwordData.isEmpty {
                return String(data: passwordData, encoding: .utf8)
            }
        }
        
        return nil
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 153
    @discardableResult
    private func keychainDeletePassword(forService serviceName: String, account: String) throws -> Bool {
        let query = newKeychainQuery(forService: serviceName, account: account)
        let status = SecItemDelete(query as CFDictionary)
        
        if status != errSecSuccess {
            throw keychainError(withCode: status)
        }
        
        return status == errSecSuccess
    }
    
    // swift-migration: original location ARTLocalDeviceStorage.m, line 162
    @discardableResult
    private func keychainSetPassword(_ password: String, forService serviceName: String, account: String) throws -> Bool {
        let passwordData = password.data(using: .utf8)!
        let searchQuery = newKeychainQuery(forService: serviceName, account: account)
        
        var status = SecItemCopyMatching(searchQuery as CFDictionary, nil)
        
        if status == errSecSuccess {
            // Item already exists, update it
            var updateQuery: [String: Any] = [:]
            updateQuery[kSecValueData as String] = passwordData
            status = SecItemUpdate(searchQuery as CFDictionary, updateQuery as CFDictionary)
        } else if status == errSecItemNotFound {
            // Item not found, create it
            var insertQuery = newKeychainQuery(forService: serviceName, account: account)
            insertQuery[kSecValueData as String] = passwordData
            status = SecItemAdd(insertQuery as CFDictionary, nil)
        }
        
        if status != errSecSuccess {
            throw keychainError(withCode: status)
        }
        
        return status == errSecSuccess
    }
}
